core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -4147,7 +4146,128 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
+    val parts = 7
+
+    val shuffleMapRdd1 = new MyRDD(sc, parts, Nil)
+    val shuffleDep1 = new ShuffleDependency(shuffleMapRdd1, new HashPartitioner(parts))
+    val shuffleMapRdd2 = new MyRDD(sc, parts, Nil)
+    val shuffleDep2 = new ShuffleDependency(shuffleMapRdd2, new HashPartitioner(parts))
+    val reduceRdd = new MyRDD(sc, parts, List(shuffleDep1, shuffleDep2),
+      tracker = mapOutputTracker)
+
+    // Submit a reduce job that depends which will create a map stage
+    submit(reduceRdd, (0 until parts).toArray)
+
+    val taskResults = taskSets(0).tasks.zipWithIndex.map {
+      case (_, idx) =>
+        (Success, makeMapStatus("host" + ('A' + idx).toChar, parts))
+    }.toSeq
+
+    val shuffleStage1 = scheduler.stageIdToStage(0).asInstanceOf[ShuffleMapStage]
+    DAGSchedulerSuite.addMergerLocs(Seq("host6", "host7", "host8"))
+    // Dummy executor added event to trigger registering of shuffle merger locations
+    // as shuffle mergers are tracked separately for test
+    runEvent(ExecutorAdded("dummy", "dummy"))
+    // Check if new shuffle merger locations are available for push or not
+    assert(mapOutputTracker.getShufflePushMergerLocations(0).size == 7)
+    assert(shuffleStage1.shuffleDep.getMergerLocs.size == 7)
+    val mergerLocsBeforeRetry = shuffleStage1.shuffleDep.getMergerLocs
+
+    // Remove MapStatus on one of the host before the stage ends to trigger
+    // a scenario where stage 0 needs to be resubmitted upon finishing all tasks.
+    // Merge finalization should be scheduled in this case.
+    for ((result, i) <- taskResults.zipWithIndex) {
+      if (i == taskSets(0).tasks.size - 1) {
+        mapOutputTracker.removeOutputsOnHost("hostA")
+      }
+      if (i < taskSets(0).tasks.size) {
+        runEvent(makeCompletionEvent(taskSets(0).tasks(i), result._1, result._2))
+      }
+    }
+
+    assert(shuffleStage1.shuffleDep.isShuffleMergeFinalized)
+
+    // Successfully completing the retry of stage 0.
+    complete(taskSets(2), taskSets(2).tasks.zipWithIndex.map {
+      case (_, idx) =>
+        (Success, makeMapStatus("host" + ('A' + idx).toChar, parts))
+    }.toSeq)
+
+    DAGSchedulerSuite.addMergerLocs(Seq("host9", "host10"))
+    // Dummy executor added event to trigger registering of shuffle merger locations
+    runEvent(ExecutorAdded("dummy1", "dummy1"))
+    assert(shuffleStage1.shuffleDep.getMergerLocs.size == 7)
+    assert(shuffleStage1.shuffleDep.isShuffleMergeFinalized)
+    complete(taskSets(1), taskSets(1).tasks.zipWithIndex.map {
+      case (_, idx) =>
+        (Success, makeMapStatus("host" + ('A' + idx).toChar, parts, 10))
+    }.toSeq)
+    val shuffleStage2 = scheduler.stageIdToStage(1).asInstanceOf[ShuffleMapStage]
+    // Shuffle merger locs should not be refreshed as the shuffle is already finalized
+    assert(mergerLocsBeforeRetry.sortBy(_.host) ===
+      shuffleStage1.shuffleDep.getMergerLocs.sortBy(_.host))
```

Comment:
We need to test for both DETERMINATE stages and INDETERMINATE stages.
We should force a new set of mergers for INDETERMINATE stage retry - In a 'real world case', this would mean adding a previously selected merger to the excludeNodes - but for the test, we can simply clear `DAGSchedulerSuite .mergerLocs` and add a different set of hosts there.